src/game/client/draw/w3dmodeldraw.cpp

```diff
@@ -1154,3 +1179,1914 @@ void W3DModelDrawModuleData::Parse_Condition_State(INI *ini, void *instance, voi
+{
+    if (((1 << wslot) & Get_W3D_Model_Draw_Module_Data()->m_projectileBoneFeedbackEnabledSlots) != 0) {
+        Do_Hide_Show_Projectile_Objects(show, count, wslot);
+    }
+}
+
+void W3DModelDraw::Update_Draw_Module_Supply_Status(int status1, int status2)
+{
+    if (status2 > 0) {
+        Get_Drawable()->Set_Model_Condition_State(MODELCONDITION_CARRYING);
+    } else {
+        Get_Drawable()->Clear_Model_Condition_State(MODELCONDITION_CARRYING);
+    }
+}
+
+void W3DModelDraw::Do_Hide_Show_Projectile_Objects(unsigned int show, unsigned int count, WeaponSlotType wslot)
+{
+    if (count >= show) {
+        int val = count - show;
+        std::vector<ModelConditionInfo::HideShowSubObjInfo> vector;
+        ModelConditionInfo::HideShowSubObjInfo info;
+
+        if (m_curState->m_weaponHideShowBoneName[wslot].Is_Empty()) {
+            for (unsigned int i = 0; i < count; i++) {
+                info.name.Format("%s%02d", m_curState->m_weaponLaunchBoneName[wslot], i + 1);
+                info.visible = (int)(i + 1) <= val;
+                vector.push_back(info);
+            }
+        } else {
+            info.name = m_curState->m_weaponHideShowBoneName[wslot];
+            info.visible = val > 0;
+            vector.push_back(info);
+        }
+
+        Do_Hide_Show_Sub_Objs(&vector);
+    } else {
+        captainslog_dbgassert(count >= show, "Someone is trying to show more projectiles than they have.");
+    }
+}
+
+void W3DModelDraw::Update_Sub_Objects()
+{
+    if (m_renderObject != nullptr && !m_subObjects.empty()) {
+        for (auto &info : m_subObjects) {
+            int index;
+            RenderObjClass *robj = m_renderObject->Get_Sub_Object_By_Name(info.name, &index);
+
+            if (robj) {
+                robj->Set_Hidden(info.visible);
+
+                const HTreeClass *tree = m_renderObject->Get_HTree();
+
+                if (tree) {
+                    Do_Hide_Show_Bone_Sub_Objs(info.visible,
+                        m_renderObject->Get_Num_Sub_Objects(),
+                        m_renderObject->Get_Sub_Object_Bone_Index(0, index),
+                        m_renderObject,
+                        tree);
+                }
+
+                robj->Release_Ref();
+            } else {
+                captainslog_dbgassert(robj,
+                    "*** ASSET ERROR: SubObject %s not found (%s)!",
+                    info.name.Str(),
+                    Get_Drawable()->Get_Template()->Get_Name().Str());
+            }
+        }
+    }
+}
+
+void W3DModelDraw::CRC_Snapshot(Xfer *xfer)
+{
+    DrawModule::CRC_Snapshot(xfer);
+}
+
+void W3DModelDraw::Xfer_Snapshot(Xfer *xfer)
+{
+    uint8_t ver = 2;
+    xfer->xferVersion(&ver, 2);
+    DrawModule::Xfer_Snapshot(xfer);
+    RecoilInfo recoilinfo;
+    for (int i = 0; i < WEAPONSLOT_COUNT; i++) {
+        unsigned char size = (unsigned char)m_weaponRecoilInfoVec[i].size();

```

Comment:
Worthwhile adding an assert checking for `m_weaponRecoilInfoVec[i].size() < 256` here?